addons/tls13-middlebox-ghack/bootstrap.js

```diff
@@ -0,0 +1,290 @@
+"use strict";
+
+let {classes: Cc, interfaces: Ci, utils: Cu, results: Cr} = Components;
+
+Cu.import("resource://gre/modules/XPCOMUtils.jsm");
+
+XPCOMUtils.defineLazyModuleGetter(this, "Preferences", "resource://gre/modules/Preferences.jsm");
+XPCOMUtils.defineLazyModuleGetter(this, "TelemetryController", "resource://gre/modules/TelemetryController.jsm");
+
+XPCOMUtils.defineLazyServiceGetter(this, "certDB", "@mozilla.org/security/x509certdb;1", "nsIX509CertDB");
+XPCOMUtils.defineLazyServiceGetter(this, "uuidGenerator", "@mozilla.org/uuid-generator;1", "nsIUUIDGenerator");
+
+const VERSION_MAX_PREF = "security.tls.version.max";
+const VERSION_FALLBACK_LIMIT_PREF = "security.tls.version.fallback-limit";
+
+const CERT_USAGE_SSL_CLIENT      = 0x0001;
+const CERT_USAGE_SSL_SERVER      = 0x0002;
+const CERT_USAGE_SSL_CA          = 0x0008;
+const CERT_USAGE_EMAIL_SIGNER    = 0x0010;
+const CERT_USAGE_EMAIL_RECIPIENT = 0x0020;
+const CERT_USAGE_OBJECT_SIGNER   = 0x0040;
+
+const REPEAT_COUNT = 5;
+
+const XHR_TIMEOUT = 10000;
+
+const TELEMETRY_PING_NAME = "tls13-middlebox-ghack";
+
+const GMAIL_URL= "https://mail.google.com/robots.txt";
+
+let configurations = [
+    {label: "google-tls13-draft-18", versionMax: 4, versionFallbackLimit: 4, website: GMAIL_URL},
+    {label: "google-tls13-exp", versionMax: 4, versionFallbackLimit: 4, website: GMAIL_URL, tlsFlags: 0x40 },    
+    {label: "google-tls12", versionMax: 3, versionFallbackLimit: 3, website: GMAIL_URL}
+];
+
+let probe_id = null;
+
+function debug(msg) {
+  console.log(msg); // eslint-disable-line no-console
+}
+
+// some fields are not available sometimes, so we have to catch the errors and return undefined.
+function getFieldValue(obj, name) {
+  try {
+    return obj[name];
+  } catch (ex) {
+    return undefined;
+  }
+}
+
+// enumerate nsIX509CertList data structure and put elements in the array
+function nsIX509CertListToArray(list) {
+  let array = [];
+
+  let iter = list.getEnumerator();
+
+  while (iter.hasMoreElements()) {
+    array.push(iter.getNext().QueryInterface(Ci.nsIX509Cert));
+  }
+
+  return array;
+}
+
+// verifies the cert using either SSL_SERVER or SSL_CA usages and extracts the chain
+// returns null in case an error occurs
+function getCertChain(cert, usage) {
+  return new Promise((resolve, reject) => {
+    certDB.asyncVerifyCertAtTime(cert, usage, 0, null, Date.now() / 1000, (aPRErrorCode, aVerifiedChain, aHasEVPolicy) => {
```

Comment:
Clocks are often wrong on the client... you might consider sending along to Telemetry what the client things the current date is, if this fails.